### PR TITLE
8272131: PhaseMacroExpand::generate_slow_arraycopy crash when clone null CallProjections.fallthrough_ioproj

### PR DESCRIPTION
**Pull request by huishi-hs**
This fix a crash in PhaseMacroExpand::generate_slow_arraycopy in following case, because arraycopy node's fallthrough_ioproj is NULL.
- arraycopy has ArrayIndexOutOfBoundsException exception
- arraycopy is before an infinite loop

Fix is not cloning and replacing fallthrough_ioproj when it is NULL.

```
    static void foo() {
        try {
            Arrays.copyOfRange(src, -1, 128);
            do {
            } while (true);
        } catch (ArrayIndexOutOfBoundsException ex) {
            count++;
        }
    }
```

New test is added.
Test: Linux X64 tier1/tier2/tier3 release/fastdebug passed.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8272131](https://bugs.openjdk.java.net/browse/JDK-8272131): PhaseMacroExpand::generate_slow_arraycopy crash when clone null CallProjections.fallthrough_ioproj

### Reviewers
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**) ⚠️ Review applies to 1c9ce9651850bf24408fcd1c9e7c0a2bb9927fd0
 * [Xin Liu](https://openjdk.java.net/census#xliu) (@navyxliu - Committer) ⚠️ Review applies to 1c9ce9651850bf24408fcd1c9e7c0a2bb9927fd0
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5054/head:pull/5054` \
`$ git checkout pull/5054`

Update a local copy of the PR: \
`$ git checkout pull/5054` \
`$ git pull https://git.openjdk.java.net/jdk pull/5054/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5054`

View PR using the GUI difftool: \
`$ git pr show -t 5054`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5054.diff">https://git.openjdk.java.net/jdk/pull/5054.diff</a>

</details>

### CODE REVIEW

**Review by neliasso — APPROVED** [review]
Looks good!

**Review by navyxliu — APPROVED** [review]
LTGM.

**Comment by huishi-hs** [comment]
@navyxliu @neliasso thanks for your review!

File: test/hotspot/jtreg/compiler/arraycopy/TestIllegalArrayCopyBeforeInfiniteLoop.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @bug 8272131
+ * @requires vm.compiler2.enabled
+ * @summary ArrayCopy with negative index before infinit loop
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/arraycopy/TestIllegalArrayCopyBeforeInfiniteLoop.java:28]: Typo `infinit` -> `infinite`

File: src/hotspot/share/opto/macroArrayCopy.cpp [hunk 2]
```diff
@@ -1095,8 +1097,14 @@ MergeMemNode* PhaseMacroExpand::generate_slow_arraycopy(ArrayCopyNode *ac,
   }
   transform_later(out_mem);
 
-  *io = _callprojs.fallthrough_ioproj->clone();
-  transform_later(*io);
+  // When src is negative and arraycopy is before an infinite loop,_callprojs.fallthrough_ioproj
+  // could be NULL. Not clone and update NULL fallthrough_ioproj.
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/macroArrayCopy.cpp:1101]: Maybe replace "Not clone" by "Skip clone".

File: test/hotspot/jtreg/compiler/arraycopy/TestIllegalArrayCopyBeforeInfiniteLoop.java (hunk 1 above)
Comment (huishi-hs) [reply test/hotspot/jtreg/compiler/arraycopy/TestIllegalArrayCopyBeforeInfiniteLoop.java]: Typo fixed.

File: src/hotspot/share/opto/macroArrayCopy.cpp (hunk 2 above)
Comment (huishi-hs) [reply src/hotspot/share/opto/macroArrayCopy.cpp]: Thanks! Updated.

**Review by TobiHartmann — APPROVED** [review]
Looks good.

**Pull request closed without merging**